### PR TITLE
[REVIEW] Added IDE changes for JNI files to have intelliSense and changes for rapids-compose [skip ci]

### PR DESCRIPTION
**Pull request by razajafri**
This PR adds the ability to use intellisense with JNI code while using rapids-compose. Thank you @trxcllnt for all your help in setting this up

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: java/pom.xml [hunk 1]
```diff
@@ -144,8 +144,11 @@
         <skipNativeCopy>false</skipNativeCopy>
         <cxx11.abi.value>ON</cxx11.abi.value>
         <cxx.flags/>
+        <cmake.export.compile.commands>OFF</cmake.export.compile.commands>
         <CUDA_STATIC_RUNTIME>OFF</CUDA_STATIC_RUNTIME>
         <PER_THREAD_DEFAULT_STREAM>OFF</PER_THREAD_DEFAULT_STREAM>
+        <project.build.directory>target/${CUDA_VERSION}</project.build.directory>
```
Comment (razajafri) [inline java/pom.xml:150]: not exactly sure if we should be modifying this variable. @tgravescs @revans2 @jlowe please comment

Comment (trxcllnt) [reply java/pom.xml:150]: Just a note, this isn't required for `rapids-compose`. The script in compose only cares about the C++ build dir, and is setting `-Dnative.build.path=` as part of the `mvn package` command.

Comment (razajafri) [reply java/pom.xml:150]: Yes, but this will help manage the java bindings in a similar way cpp build is managed. 
I still need to add the branch name to this property

File: java/pom.xml [hunk 2]
```diff
@@ -234,6 +246,7 @@
     </profiles>
 
     <build>
+	    <directory>${project.build.directory}</directory>
```
Comment (revans2) [inline java/pom.xml:249]: nit: indentation appears to be off here.

File: java/pom.xml [hunk 3]
```diff
@@ -144,11 +144,23 @@
         <skipNativeCopy>false</skipNativeCopy>
         <cxx11.abi.value>ON</cxx11.abi.value>
         <cxx.flags/>
+        <CMAKE_EXPORT_COMPILE_COMMANDS>OFF</CMAKE_EXPORT_COMPILE_COMMANDS>
         <CUDA_STATIC_RUNTIME>OFF</CUDA_STATIC_RUNTIME>
         <PER_THREAD_DEFAULT_STREAM>OFF</PER_THREAD_DEFAULT_STREAM>
+        <project.build.directory>target</project.build.directory>
```
Comment (revans2) [inline java/pom.xml:150]: Why are we setting the project build directory at all for the default profile?

File: java/pom.xml [hunk 4]
```diff
@@ -144,11 +144,23 @@
...
     </properties>
 
     <profiles>
+        <profile> 
+            <id>managed-release-directory</id>
+            <activation> 
+                <activeByDefault>false</activeByDefault>
+            </activation>
+            <properties>
+                <project.build.directory>target/cuda-${CUDA_VERSION}/${git.branch}/release</project.build.directory>
```
Comment (revans2) [inline java/pom.xml:161]: Where did `CUDA_VERSION` and `git.branch` come from?

File: java/pom.xml (hunk 2 above)
Comment (revans2) [reply java/pom.xml:249]: Why are we setting this at all?

File: java/pom.xml (hunk 4 above)
Comment (razajafri) [reply java/pom.xml:161]: They will be set in the docker for rapids-compose.

File: java/pom.xml (hunk 2 above)
Comment (razajafri) [reply java/pom.xml:249]: Under normal flow it will be the default directory but when we are in compose we want the user to be able to work on different branches and with different cuda versions. By setting this value we are able to control where the artifacts will be copied. Is there a better way to do this?

Comment (razajafri) [reply java/pom.xml:249]: if there's a way to conditionally add the `<directory>` tag, that will be ideal

File: java/pom.xml (hunk 4 above)
Comment (jlowe) [reply java/pom.xml:161]: Is rapids-compose now a requirement to build this?

Comment (razajafri) [reply java/pom.xml:161]: No, here's what I am trying to achieve. when you set the `CUDA_VERSION` and `git.branch` variables and call this profile. It will put the artifacts in `target/cuda-${CUDA_VERSION}/${git.branch}/release`. 
Ideally I didn't want to change the pom at all but the `project.build.output` can't be set from the command line. 

I think it will be better if I focus this PR only on the rapids-compose change and undo the change to the output directory.

Comment (trxcllnt) [reply java/pom.xml:161]: It shouldn't be, and imo we should avoid adding references to values exclusive to the rapids-compose environment.

The intention is to prefix the C++ build directories with `cuda_version/branch_name/build_type`, because it helps improve (or avoid) costly rebuild times changing those values. We can achieve this in compose by setting `-Dnative.build.path` in the `mvn package` command.

Since Java rebuild times are fast compared to C++, I think it's overkill to prefix the entire `target` directory of Java build artifacts with those values.

File: java/src/main/native/CMakeLists.txt [hunk 5]
```diff
@@ -146,11 +151,13 @@ endif (CUDF_INCLUDE AND CUDF_LIBRARY)
 
 find_path(RMM_INCLUDE "rmm"
           HINTS "$ENV{RMM_ROOT}/include"
+                "$ENV{RMM_HOME}/include"
                 "$ENV{CONDA_PREFIX}/include/rmm"
                 "$ENV{CONDA_PREFIX}/include")
 
 find_library(RMM_LIBRARY "rmm"
              HINTS "$ENV{RMM_ROOT}/lib"
```
Comment (razajafri) [inline java/src/main/native/CMakeLists.txt:159]: Is this necessary? If others are OK I would like to remove this

File: java/src/main/native/CMakeLists.txt [hunk 6]
```diff
@@ -146,11 +151,13 @@ endif (CUDF_INCLUDE AND CUDF_LIBRARY)
 
 find_path(RMM_INCLUDE "rmm"
           HINTS "$ENV{RMM_ROOT}/include"
```
Comment (razajafri) [inline java/src/main/native/CMakeLists.txt:153]: Is this necessary? If others are OK I would like to remove this

File: java/src/main/native/CMakeLists.txt [hunk 7]
```diff
@@ -129,6 +133,7 @@ set(CUDF_INCLUDE "${PROJECT_SOURCE_DIR}/../../../../cpp/include"
 
 find_library(CUDF_LIBRARY "cudf"
     HINTS "$ENV{CUDF_ROOT}/lib"
```
Comment (razajafri) [inline java/src/main/native/CMakeLists.txt:135]: Is this necessary? If others are OK I would like to remove this. I don't think we need this to build on the host directly.

File: java/pom.xml (hunk 3 above)
Comment (razajafri) [reply java/pom.xml]: @revans2 I have removed the change to the project build directory. PTAL

File: java/src/main/native/CMakeLists.txt (hunk 6 above)
Comment (jlowe) [reply java/src/main/native/CMakeLists.txt:153]: This allows users to override the prefix where RMM has been installed via the `RMM_ROOT` environment variable, just as they can with cudf via `CUDF_ROOT`.

File: java/src/main/native/CMakeLists.txt (hunk 5 above)
Comment (jlowe) [reply java/src/main/native/CMakeLists.txt:159]: In the RMM case, this is already planned to be removed by #5632.  Let's keep this PR focused on the necessary changes.

File: java/src/main/native/CMakeLists.txt (hunk 7 above)
Comment (revans2) [reply java/src/main/native/CMakeLists.txt:135]: It follows the conventions of the other libraries. and is where we would find libcudf if it were installed instead of just being built.

**Comment by razajafri** [comment]
@revans2 @jlowe thanks for the review, is there anything else you want me to change?

File: java/src/main/native/CMakeLists.txt [hunk 8]
```diff
@@ -146,11 +151,13 @@ endif (CUDF_INCLUDE AND CUDF_LIBRARY)
 
 find_path(RMM_INCLUDE "rmm"
           HINTS "$ENV{RMM_ROOT}/include"
+                "$ENV{RMM_HOME}/include"
```
Comment (jlowe) [inline java/src/main/native/CMakeLists.txt:154]: What's going on with `RMM_HOME`?  I don't see this being used in the cpp CMakeLists.txt, and given there's already an `RMM_ROOT` used by both this and the cpp CMakeLists.txt, I would not expect there to be a need for `RMM_HOME` especially since we don't see it in the cpp build.

File: java/src/main/native/CMakeLists.txt [hunk 9]
```diff
@@ -129,6 +133,7 @@ set(CUDF_INCLUDE "${PROJECT_SOURCE_DIR}/../../../../cpp/include"
 
 find_library(CUDF_LIBRARY "cudf"
     HINTS "$ENV{CUDF_ROOT}/lib"
+          "$ENV{CUDF_ROOT}"
```
Comment (jlowe) [inline java/src/main/native/CMakeLists.txt:136]: `CUDF_ROOT` is supposed to be set to the install prefix of cudf, so I don't think we want to be looking both in `CUDF_ROOT` _and_ `CUDF_ROOT/lib`.

**Comment by trxcllnt** [comment]
tl;dr: these two commits are the minimal set of changes I've tested that work in compose:
1. [`pom.xml`](https://github.com/trxcllnt/cudf/commit/c523b24d5fa1909295e559ff700fb7edee21a68c)
2. [`CMakeLists.txt`](https://github.com/trxcllnt/cudf/commit/f6ced5203bb2c778c10d4c5c07d593c10f543d12)

@jlowe The changes to the `CMakeLists.txt` are to support source-building the RAPIDS projects against each other (without installing). We don't have an "install" location in the traditional sense, because we only run `make` and not `make install`. 

The loose pattern in our CI and `build.sh` scripts is:
* `<PROJ>_HOME` env vars point at the repository root
* `<PROJ>_ROOT` env vars point at the install location, or occasionally the C++ build dir.

I've replicated this pattern in `rapids-compose` for consistency:
```bash
RMM_HOME=rapids/rmm
RMM_ROOT=rapids/rmm/build/release
CUDF_HOME=rapids/cudf
CUDF_ROOT=rapids/cudf/cpp/build/release
```

With CMake I'd override the paths on the command line like this:
```
cmake -D RMM_INCLUDE=$RMM_HOME/include \
      -D RMM_LIBRARY=$RMM_ROOT/librmm.so \
      -D CUDF_INCLUDE=$CUDF_HOME/cpp/include \
      -D CUDF_LIBRARY=$CUDF_ROOT/libcudf.so \
    ..
```

But @razajafri and I couldn't find a good way to use `mvn package` and pass those arguments onto CMake. We can modify `pom.xml` to forward on each CLI argument (like `-DARROW_STATIC_LIB`), but there isn't a good default for each of those vars if you invoke `mvn package` standalone. Defaulting to nothing causes CMake to think the empty string should be the location.

So the alternative is to have the `CMakeLists.txt` look in additional locations via environment variables. Neither RMM nor cuDF have `lib` or `include` directories in their C++ build dirs, which is why we set `RMM_INCLUDE` and `CUDF_INCLUDE` to their source locations.

File: java/src/main/native/CMakeLists.txt (hunk 8 above)
Comment (razajafri) [reply java/src/main/native/CMakeLists.txt:154]: cpp CMakeLists.txt doesn't have `RMM_HOME` as a hint and the reason its working when we build it in rapids-compose is because we can directly pass the cmake variables but unfortunately there is no way that I and @trxcllnt could find to override those variables conditionally in `antrun` plugin in maven. so when we don't pass any variables to maven they are passed as empty strings which causes the cmake to assign those values to the `RMM_INCLUDE` etc causing build failures. 

One thing that I completely forgot to mention in our chat offline, and as Paul has pointed out in his comment, is that in rapids-compose we are doing source building so, there is no install. cudf and rmm both don't copy their headers to the build directory and neither have a lib folder in the build directory. 

I hope this justifies this change

**Comment by jlowe** [comment]
Thanks for the clarification, @trxcllnt.  That helps explain why `$RMM_HOME` was showing up in the JNI CMakeLists.txt but not the cpp version.

**Comment by razajafri** [comment]
@jlowe do you have any other comments?

File: java/src/main/native/CMakeLists.txt [hunk 10]
```diff
@@ -121,14 +121,18 @@ find_path(LIBCUDACXX_INCLUDE "simt"
...
+
 ###################################################################################################
 # - CUDF ------------------------------------------------------------------------------------------
 
 set(CUDF_INCLUDE "${PROJECT_SOURCE_DIR}/../../../../cpp/include"
                  "${PROJECT_SOURCE_DIR}/../../../../cpp/src/")
 
 find_library(CUDF_LIBRARY "cudf"
-    HINTS "$ENV{CUDF_ROOT}/lib"
+    HINTS "$ENV{CUDF_ROOT}"
```
Comment (jlowe) [inline java/src/main/native/CMakeLists.txt:135]: $CUDF_ROOT/lib should remain in case the user set this environment variable to be the install prefix.
